sdk/core/Azure.Core/src/Shared/OperationInternal.cs

```diff
@@ -0,0 +1,207 @@
+
+        public string OperationTypeName { get; set; }
+
+        public TimeSpan DefaultPollingInterval { get; set; }
+
+        private ClientDiagnostics ClientDiagnostics { get; }
+
+        private RequestFailedException OperationFailedException
+        {
+            get => _operationFailedException;
+            set
+            {
+                if (value is null)
+                {
+                    throw new ArgumentNullException(nameof(value));
+                }
+
+                _operationFailedException = value;
+            }
+        }
+
+        public async ValueTask<Response> UpdateStatusAsync(CancellationToken cancellationToken) =>
+            await UpdateStatusAsync(async: true, cancellationToken).ConfigureAwait(false);
+
+        public Response UpdateStatus(CancellationToken cancellationToken) =>
+            UpdateStatusAsync(async: false, cancellationToken).EnsureCompleted();
+
+        public async ValueTask<Response<TResult>> WaitForCompletionAsync(CancellationToken cancellationToken) =>
+            await WaitForCompletionAsync(DefaultPollingInterval, cancellationToken).ConfigureAwait(false);
+
+        public async ValueTask<Response<TResult>> WaitForCompletionAsync(TimeSpan pollingInterval, CancellationToken cancellationToken)
+        {
+            while (true)
+            {
+                Response response = await UpdateStatusAsync(cancellationToken).ConfigureAwait(false);
+
+                if (HasCompleted)
+                {
+                    return Response.FromValue(Value, response);
+                }
+
+                var serverDelay = response.Headers.TryGetValue(RetryAfterHeaderName, out string retryAfterValue)
+                    && int.TryParse(retryAfterValue, out int serverDelayInSeconds)
+                    ? TimeSpan.FromSeconds(serverDelayInSeconds) : TimeSpan.Zero;
+
+                var delay = serverDelay > pollingInterval
+                    ? serverDelay : pollingInterval;
+
+                await Task.Delay(delay, cancellationToken).ConfigureAwait(false);
+            }
+        }
+
+        private async ValueTask<Response> UpdateStatusAsync(bool async, CancellationToken cancellationToken)
+        {
+            using DiagnosticScope scope = ClientDiagnostics.CreateScope($"{OperationTypeName}.UpdateStatus");
+
+            foreach (KeyValuePair<string, string> attribute in ScopeAttributes)
+            {
+                scope.AddAttribute(attribute.Key, attribute.Value);
+            }
+
+            scope.Start();
+
+            try
+            {
+                return await UpdateStateAsync(async, cancellationToken).ConfigureAwait(false);
+            }
+            catch (RequestFailedException e)
+            {
+                scope.Failed(e);
+                throw;
+            }
+            catch (Exception e)
+            {
+                var requestFailedException = new RequestFailedException("Unexpected failure.", e);
```

Comment:
I think we should just flow the exception out.

Reply:
Flowing the exception out is reasonable, and we had some operations doing it. We also have some operations that encapsulate it in a `RequestFailedException` (Key Vault, Synapse). I believe the reasoning behind it is that we can still say that "the request failed", even though we didn't make an actual service call.

We could, on the other hand, flow the exception out like you suggested and treat the operations I mentioned as special cases. We could make workarounds for them:
```cs
public override Response UpdateStatus(CancellationToken cancellationToken = default)
{
    try
    {
        return _operationInternal.UpdateStatus(cancellationToken);
    }
    catch (Exception e)
    {
        // wrap exception in a RequestFailedException and throw to keep original behavior
    }
}
```
What do you think?

Reply:
Yeah, let's special case this behavior in the KV itself but also ask them if this is something they are willing to change.

Reply:
The Key Vault operation we're updating in this PR was not affected by it, but the KV client has some operations with this behavior to be updated later.

Synapse has been updated:
https://github.com/Azure/azure-sdk-for-net/blob/c24a61f1bf175f9723bf5868fc8141c96ecb088e/sdk/synapse/Azure.Analytics.Synapse.Spark/src/Customization/SparkSessionOperation.cs#L62-L72

We're doing this in `UpdateStatus`, `UpdateStatusAsync`, and both `WaitForCompletionAsync` overloads.